deno/lib/types.ts

```diff
@@ -359,6 +360,30 @@ export abstract class ZodType<
     return returnType;
   }
 
+  refineWith<NewOut, This extends this>(
+    returnType: ZodType<NewOut>
+  ): This extends ZodEffects<infer T, any>
+    ? ZodEffects<T, NewOut>
+    : ZodEffects<This, NewOut> {
+    let parsed: ReturnType<typeof returnType.safeParse>;
```

Comment:
Actually I suppose there's a concurrency bug here due to this being in the scope of the schema rather than the specific parsing invocation of it. So `parsed` would have to be stashed somewhere within the context.